File: frontend/src/js/syntaxHighlighting.js
/**
 * Syntax highlighting service that uses the backend API for comprehensive language support
 */

import config from './config.js';

class SyntaxHighlightingService {
    constructor() {
        this.cache = new Map(); // Cache highlighted code to avoid redundant API calls
        this.supportedLanguages = new Set(); // Cache of supported languages
        this.apiBase = config.apiBaseUrl;
    }

    /**
     * Highlight code using the backend API
     * @param {string} code - The source code to highlight
     * @param {string} language - The programming language
     * @returns {Promise<string>} - HTML with syntax highlighting
     */
    async highlightCode(code, language) {
        if (!code || !code.trim()) {
            return this.escapeHtml(code);
        }

        if (!language) {
            return this.escapeHtml(code);
        }

        // Create cache key
        const cacheKey = `${language}:${this.hashCode(code)}`;

        // Check cache first
        if (this.cache.has(cacheKey)) {
            return this.cache.get(cacheKey);
        }

        try {
            const response = await fetch(`${this.apiBase}/highlight/highlight`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    code: code,
                    language: language
                })
            });

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}`);
            }

            const result = await response.json();

            // Cache the result
            this.cache.set(cacheKey, result.highlighted_code);

            // Update supported languages cache
            if (result.is_supported) {
                this.supportedLanguages.add(language.toLowerCase());
            }

            return result.highlighted_code;

        } catch (error) {
            console.warn(`Failed to highlight code for language '${language}':`, error);
            // Fallback to escaped HTML
            const escaped = this.escapeHtml(code);
            this.cache.set(cacheKey, escaped);
            return escaped;
        }
    }

    /**
     * Check if a language is supported
     * @param {string} language - The programming language
     * @returns {Promise<boolean>} - Whether the language is supported
     */
    async isLanguageSupported(language) {
        if (!language) return false;

        const langLower = language.toLowerCase();

        // Check cache first
        if (this.supportedLanguages.has(langLower)) {
            return true;
        }

        try {
            const response = await fetch(`${this.apiBase}/highlight/languages/${encodeURIComponent(language)}/check`);

            if (!response.ok) {
                return false;
            }

            const result = await response.json();

            if (result.is_supported) {
                this.supportedLanguages.add(langLower);
            }

            return result.is_supported;

        } catch (error) {
            console.warn(`Failed to check language support for '${language}':`, error);
            return false;
        }
    }

    /**
     * Get all available languages
     * @returns {Promise<Object>} - Object mapping language aliases to full names
     */
    async getAvailableLanguages() {
        try {
            const response = await fetch(`${this.apiBase}/highlight/languages`);

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}`);
            }

            const result = await response.json();
            return result.languages;

        } catch (error) {
            console.warn('Failed to get available languages:', error);
            return {};
        }
    }

    /**
     * Clear the highlighting cache
     */
    clearCache() {
        this.cache.clear();
        this.supportedLanguages.clear();
    }

    /**
     * Escape HTML characters
     * @param {string} text - Text to escape
     * @returns {string} - Escaped text
     */
    escapeHtml(text) {
        const div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    }

    /**
     * Simple hash function for cache keys
     * @param {string} str - String to hash
     * @returns {number} - Hash value
     */
    hashCode(str) {
        let hash = 0;
        if (str.length === 0) return hash;
        for (let i = 0; i < str.length; i++) {
            const char = str.charCodeAt(i);
            hash = ((hash << 5) - hash) + char;
            hash = hash & hash; // Convert to 32-bit integer
        }
        return hash;
    }
}

// Export singleton instance
export default new SyntaxHighlightingService();

File: backend/app/services/syntax_highlighting.py
"""
Syntax highlighting service using Pygments.
Provides comprehensive syntax highlighting for code blocks with support
for many programming languages.
"""
import html
from typing import Any, Dict, Optional

from pygments import highlight
from pygments.formatters import HtmlFormatter
from pygments.lexers import get_all_lexers, get_lexer_by_name
from pygments.util import ClassNotFound


class SyntaxHighlightingService:
    """Service for syntax highlighting code blocks using Pygments."""

    def __init__(self) -> None:
        """Initialize the syntax highlighting service."""
        self.formatter: HtmlFormatter = HtmlFormatter(
            nowrap=True,
            cssclass="",
            classprefix="token ",
        )
        self._available_languages: Optional[Dict[str, str]] = None

    def get_available_languages(self) -> Dict[str, str]:
        """Get all available languages that can be highlighted."""
        if self._available_languages is None:
            self._available_languages = {}
            for lexer in get_all_lexers():
                name = lexer[0]
                aliases = lexer[1]
                if aliases:
                    for alias in aliases:
                        self._available_languages[alias.lower()] = name
        return self._available_languages

    def highlight_code(self, code: str, language: Optional[str] = None) -> str:
        """
        Highlight code using Pygments.
        Args:
            code: The source code to highlight
            language: The programming language (optional)
        Returns:
            HTML string with syntax highlighting
        """
        if not code or not code.strip():
            return html.escape(code)
        if not language:
            return html.escape(code)
        try:
            lexer = get_lexer_by_name(language.lower(), stripnl=False)
            highlighted = highlight(code, lexer, self.formatter)
            return str(highlighted)
        except ClassNotFound:
            return html.escape(code)
        except Exception as e:
            print(f"Error highlighting code: {e}")
            return html.escape(code)

    def is_language_supported(self, language: str) -> bool:
        """Check if a language is supported for highlighting."""
        if not language:
            return False
        try:
            get_lexer_by_name(language.lower())
            return True
        except ClassNotFound:
            return False

    def get_language_info(self, language: str) -> Optional[Dict[str, Any]]:
        """Get information about a specific language."""
        try:
            lexer = get_lexer_by_name(language.lower())
            return {
                "name": lexer.name,
                "aliases": lexer.aliases,
                "filenames": lexer.filenames,
                "mimetypes": lexer.mimetypes,
            }
        except ClassNotFound:
            return None


# Global instance
syntax_highlighter = SyntaxHighlightingService()
